test/prototype/test_low_bit_optim.py

```diff
@@ -156,6 +166,81 @@ def test_optim_fp8_smoke(self, optim_name, device):
         optim.zero_grad()
 
 
+class TestFSDP2(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @pytest.mark.skipif(not TORCH_VERSION_AFTER_2_4, reason="torch >= 2.4 required")
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp2(self):
+        self.run_subtests(
+            {
+                "enable_activation_checkpointing": [False, True],
+                "offload_policy": [
+                    OffloadPolicy(),
+                    # CPUOffloadPolicy(pin_memory=True),
+                    # CPUOffloadPolicy(pin_memory=False),
+                ],
+            },
+            self._test_fsdp2,
+        )
+
+    def _test_fsdp2(self, enable_activation_checkpointing, offload_policy):
+        from torch.testing._internal.distributed._tensor.common_dtensor import (
+            ModelArgs,
+            Transformer,
+            TransformerBlock,
+        )
+
+        batch_size = 3
+        vocab_size = 1024
+        seq_len = 64
+        model_args = ModelArgs(
+            n_layers=3,
+            n_heads=4,
+            dim=1024,
+            vocab_size=vocab_size,
+            max_seq_len=seq_len,
+            dropout_p=0,
+        )
+        torch.manual_seed(42)
+        with torch.device("cuda"):
+            base_model = Transformer(model_args)
+        if enable_activation_checkpointing:
+            apply_activation_checkpointing(base_model, auto_wrap_policy=ModuleWrapPolicy({TransformerBlock}))
+        base_optim = low_bit_optim.Adam8bit(base_model.parameters(), lr=1e-2)
+
+        fsdp_kwargs = {"offload_policy": offload_policy}
+        fsdp_model = copy.deepcopy(base_model)
+        for m in fsdp_model.modules():
+            if enable_activation_checkpointing:
+                if isinstance(m, CheckpointWrapper):
+                    fully_shard(m, **fsdp_kwargs)
+            else:
+                if isinstance(m, TransformerBlock):
+                    fully_shard(m, **fsdp_kwargs)
+        fully_shard(fsdp_model, **fsdp_kwargs)
+        fsdp_optim = low_bit_optim.Adam8bit(fsdp_model.parameters(), lr=1e-2)
```

Comment:
Is `.grad` converted from high precision (fp32 or bf16) to 8bit inside `Adam8bit` ?  Asking because FSDP2 seems to be performing gradient reduction in high precision base on `fsdp_kwargs`

Reply:
Grad is in original precision. To summarize
1. optim state is dequant to FP32 (maybe I need to change it to param or grad dtype? haven't test full BF16 training)
2. adam step as usual
3. optim state is quantized back to 8-bit

Reply:
thanks for explaining. that means FSDP2 logic during fwd/bwd remains the same (doing high precision gradient reduction) no matter optimizer is in high precision or low precision. The above dicussion around DTensor is the core then